src/csexp.ml

```diff
@@ -58,36 +59,41 @@ module Make (Sexp : Sexp) = struct
         | Ok s -> return @@ Ok (Atom s)
         | Error e -> return @@ Error e )
       | Ok c -> return @@ invalid_character c
-      | Error e -> return @@ Error e
 
     let rec parse_many depth input acc =
       Input.read_char input >>= function
       | Ok '(' -> (
         parse_many (depth + 1) input [] >>= function
-        | Ok sexps -> parse_many (depth + 1) input @@ (List sexps :: acc)
+        | Ok sexps -> parse_many depth input @@ (List sexps :: acc)
         | e -> return e )
       | Ok ')' ->
-        if depth = 0 then
-          return @@ missing_left_parenthesis ()
-        else
-          return @@ Ok (List.rev acc)
+        return
+          ( if depth = 0 then
+            missing_left_parenthesis ()
+          else
+            Ok (List.rev acc) )
       | Ok c when '0' <= c && c <= '9' -> (
         parse_atom input (int_of_digit c) >>= function
         | Ok sexp -> parse_many depth input (sexp :: acc)
         | Error e -> return @@ Error e )
       | Ok c -> return @@ invalid_character c
-      | Error e -> return @@ Error e
+      | Error e ->
+        return
```

Comment:
This doesn't feel right. Now that I look at the API again, I understand why this is like this. Errors are string so there isn't much we can do. I must have overlooked this when reviewing the PR adding the functor. We clearly should have added tests at this point...

Well not sure we can do much better now, so let's go with it. But we'll need to prepare a 2.0 with a new API.

Reply:
I also feel the current API isn't quite right. I was planning to raise an issue to make the `input` functions more like:

```
type 'a or_eof = Value of 'a | Eof
val input_char : t -> (char or_eof, string) Result.t
val input_ string : t -> int -> (string or_eof, string) Result.t
```

I'm not sure of the concrete API here, but I think it would be less confusing if we separated the enf of the stream from an actual error in the signature.

Reply:
agreed. In fact, we shouldn't even try to convert IO errors to strings. That seems wrong to me. We should let the IO monad deal with IO errors the way it wants to.

Reply:
So that would be:

```ocaml
val read_char : t -> char option Monad.t
val read_string : t -> string option Monad.t
```

I just gave this a try. Submitting a PR now.